File: examples/python/contact_book.py
#!/usr/bin/env python3
# Copyright (c) Andreas Flakstad and Vev contributors
# SPDX-License-Identifier: EPL-2.0

from __future__ import annotations

import pathlib
import sys
import tempfile

ROOT = pathlib.Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / "clients" / "python"))

import vev


SCHEMA = """
[[:db/add 10 :db/ident :contact/email]
 [:db/add 10 :db/unique :db.unique/identity]
 [:db/add 11 :db/ident :contact/knows]
 [:db/add 11 :db/valueType :db.type/ref]]
"""

DATA = """
[{:db/id 1
  :contact/name "Ada Lovelace"
  :contact/email "ada@example.com"
  :contact/company "Analytical Engines"}
 {:db/id 2
  :contact/name "Grace Hopper"
  :contact/email "grace@example.com"
  :contact/company "Compilers"}
 [:db/add 1 :contact/knows 2]]
"""

KATHERINE = """
[{:db/id 3
  :contact/name "Katherine Johnson"
  :contact/email "katherine@example.com"
  :contact/company "NASA"}]
"""


def remove_store(path: pathlib.Path) -> None:
    for suffix in ("", "-wal", "-shm"):
        try:
            pathlib.Path(f"{path}{suffix}").unlink()
        except FileNotFoundError:
            pass


def seed(conn: vev.Connection | vev.DurableConnection) -> None:
    conn.transact(SCHEMA)
    conn.transact(DATA)


def contact_names(source: vev.Connection | vev.DurableConnection | vev.DB) -> list[str]:
    rows = vev.q(
        '[:find ?name :where [?e :contact/name ?name]]',
        source,
    )
    return sorted(row[0] for row in rows)


def assert_base_contacts(db: vev.DB) -> None:
    if contact_names(db) != ["Ada Lovelace", "Grace Hopper"]:
        raise RuntimeError("unexpected contact names")

    rows = vev.q(
        """
        [:find ?name
         :in $ ?email
         :where [?e :contact/email ?email]
                [?e :contact/name ?name]]
        """,
        db,
        '["ada@example.com"]',
    )
    if rows != [["Ada Lovelace"]]:
        raise RuntimeError("lookup query returned the wrong contact")

    profile = db.pull(
        "[:contact/name :contact/email {:contact/knows [:contact/name]}]",
        vev.Entity(1),
    )
    if profile.get(":contact/name") != "Ada Lovelace":
        raise RuntimeError("pull returned the wrong contact")
    if profile.get(":contact/knows", {}).get(":contact/name") != "Grace Hopper":
        raise RuntimeError("pull returned the wrong related contact")


def run_in_memory() -> None:
    with vev.create_conn() as conn:
        seed(conn)
        with conn.db() as db:
            assert_base_contacts(db)
            with db.db_with(KATHERINE) as next_db:
                if "Katherine Johnson" not in contact_names(next_db):
                    raise RuntimeError("db_with did not add the hypothetical contact")
                if "Katherine Johnson" in contact_names(db):
                    raise RuntimeError("db_with mutated the original DB value")


def run_durable(path: pathlib.Path) -> None:
    remove_store(path)

    with vev.connect(path) as conn:
        seed(conn)
        with conn.db() as db:
            assert_base_contacts(db)

    with vev.connect(path) as conn:
        if contact_names(conn) != ["Ada Lovelace", "Grace Hopper"]:
            raise RuntimeError("reopened durable store did not preserve contacts")
        conn.transact(KATHERINE)
        if "Katherine Johnson" not in contact_names(conn):
            raise RuntimeError("durable transaction did not update the connection DB")

    with vev.connect(path) as conn:
        if "Katherine Johnson" not in contact_names(conn):
            raise RuntimeError("reopened durable store did not preserve the update")

    remove_store(path)


def main(argv: list[str]) -> int:
    if len(argv) > 1:
        durable_path = pathlib.Path(argv[1])
    else:
        durable_path = pathlib.Path(tempfile.gettempdir()) / "vev-contact-book.vev"

    run_in_memory()
    run_durable(durable_path)
    print("contact-book: ok")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: examples/clojure/contact_book.clj
;; Copyright (c) Andreas Flakstad and Vev contributors
;; SPDX-License-Identifier: EPL-2.0

(ns contact-book
  (:require [clojure.java.io :as io]
            [vev.core :as d]))

(def schema
  [[:db/add 10 :db/ident :contact/email]
   [:db/add 10 :db/unique :db.unique/identity]
   [:db/add 11 :db/ident :contact/knows]
   [:db/add 11 :db/valueType :db.type/ref]])

(def contacts
  [{:db/id 1
    :contact/name "Ada Lovelace"
    :contact/email "ada@example.com"
    :contact/company "Analytical Engines"}
   {:db/id 2
    :contact/name "Grace Hopper"
    :contact/email "grace@example.com"
    :contact/company "Compilers"}
   [:db/add 1 :contact/knows 2]])

(def katherine
  [{:db/id 3
    :contact/name "Katherine Johnson"
    :contact/email "katherine@example.com"
    :contact/company "NASA"}])

(def names-query
  '[:find ?name
    :where [?e :contact/name ?name]])

(def email-query
  '[:find ?name
    :in $ ?email
    :where [?e :contact/email ?email]
           [?e :contact/name ?name]])

(def contact-pattern
  [:contact/name :contact/email {:contact/knows [:contact/name]}])

(defn remove-store! [path]
  (doseq [suffix ["" "-wal" "-shm"]]
    (io/delete-file (str path suffix) true)))

(defn seed! [conn]
  (d/transact conn schema)
  (d/transact conn contacts))

(defn contact-names [source]
  (->> (d/q names-query source)
       (map first)
       sort
       vec))

(defn assert-base-contacts! [db]
  (assert (= ["Ada Lovelace" "Grace Hopper"] (contact-names db)))
  (assert (= #{["Ada Lovelace"]}
             (d/q email-query db "ada@example.com")))
  (let [profile (d/pull db contact-pattern 1)]
    (assert (= "Ada Lovelace" (:contact/name profile)))
    (assert (= "Grace Hopper" (get-in profile [:contact/knows :contact/name])))))

(defn run-in-memory! []
  (with-open [conn (d/create-conn)]
    (seed! conn)
    (let [db (d/db conn)
          next-db (d/db-with db katherine)]
      (assert-base-contacts! db)
      (assert (some #{"Katherine Johnson"} (contact-names next-db)))
      (assert (not-any? #{"Katherine Johnson"} (contact-names db))))))

(defn run-durable! [path]
  (remove-store! path)
  (with-open [conn (d/connect path)]
    (seed! conn)
    (let [db (d/db conn)]
      (assert-base-contacts! db)))
  (with-open [conn (d/connect path)]
    (assert (= ["Ada Lovelace" "Grace Hopper"] (contact-names conn)))
    (d/transact conn katherine)
    (assert (some #{"Katherine Johnson"} (contact-names conn))))
  (with-open [conn (d/connect path)]
    (assert (some #{"Katherine Johnson"} (contact-names conn))))
  (remove-store! path))

(defn -main [& [path]]
  (let [store (or path
                  (str (System/getProperty "java.io.tmpdir")
                       "/vev-contact-book-clojure.vev"))]
    (run-in-memory!)
    (run-durable! store)
    (println "contact-book-clojure: ok")))
